### DISCUSSION TITLE
Beamforming and Diversity

### DISCUSSION THREAD

**Discussion by colinrenard** [NVlabs/sionna#676] [Q&A] [answered]
Hello everyone, 

I am trying to implement a MU-MIMO downlink scenario and I intended to use beamforming combined with diversity.
The scenario would involve 4 users (2 antennas per user), and 8 antennas at the base station. Since I want to create 4 beams, I set the num_streams_per_tx parameter to 4, but this results in a shape error with the ZFPrecoder instance. To my understanding, num_streams_per_tx = 8 would result in multiplexing instead of diversity.

Here is the code.: 
[OFDM_Beamforming.zip](https://github.com/user-attachments/files/17906212/OFDM_Beamforming.zip)

Is there a workaround to be able to implement diversity and multiplexing simultaneously? 
Thanks in advance!

**Comment by jhoydis** [comment, marked as answer]
Hi @colinrenard,

Without looking into your code, the issue is probably the following:

The ZFPrecoder sees an 8x8  channel matrix which it tries to invert. It essentially sends an independent stream to each of the receive antennas in the system.

However, you would like to have some form of receive combining at the users, e.g., simply add the received signal at both antennas together for diversity. This receive combining should be part of the "effective 4x8 channel" that the ZF preceder tries to pseudo-invert.

If we denote by $\mathbf{H}\in\mathbb{C}^{8\times 8}$ the full channel matrix  and by $\mathbf{G} \in \mathbb{C}^{4\times 8}$ the receive combining matrix, then the effective channel you should give to the precoder is $\mathbf{H}_\text{eff}=\mathbf{G}\mathbf{H}$.

Hope that helps.

**Reply by colinrenard** [reply]
Hi, 

Thanks for the quick reply and the suggestion!
